qubesusbproxy/tests.py

```diff
@@ -524,76 +667,521 @@ def test_061_auto_attach_on_reconnect(self):
+        return False
+
+    def __str__(self):
+        return self.name
+
+
+def get_qdb(attachment=None):
+    result = {
+        "/qubes-usb-devices/1-1/desc": b"1a0a:badd USB-IF Test\x20Device",
+        "/qubes-usb-devices/1-1/interfaces": b":ffff00:020600:0a0000:",
+        "/qubes-usb-devices/1-1/usb-ver": b"2",
+        "/qubes-usb-devices/1-2/desc": b"1a0a:badd USB-IF Test\x20Device\x202",
+        "/qubes-usb-devices/1-2/interfaces": b":0acafe:",
+        "/qubes-usb-devices/1-2/usb-ver": b"3",
+    }
+    if attachment:
+        result["/qubes-usb-devices/1-1/connected-to"] = attachment.encode()
+    return result
+
+
+class TC_30_USBProxy_core3(qubes.tests.QubesTestCase):
+    # noinspection PyAttributeOutsideInit
+    def setUp(self):
+        super().setUp()
+        self.ext = qubesusbproxy.core3ext.USBDeviceExtension()
+
+    @staticmethod
+    def added_assign_setup(attachment=None):
+        back_vm = TestVM(qdb=get_qdb(attachment), name="sys-usb")
+        front = TestVM({}, name="front-vm")
+        dom0 = TestVM({}, name="dom0")
+        back_vm.app.domains["sys-usb"] = back_vm
+        back_vm.app.domains["front-vm"] = front
+        back_vm.app.domains[0] = dom0
+        back_vm.app.domains["dom0"] = dom0
+        front.app = back_vm.app
+        dom0.app = back_vm.app
+
+        back_vm.app.vmm.configure_mock(**{"offline_mode": False})
+        fire_event_async = mock.Mock()
+        front.fire_event_async = fire_event_async
+
+        back_vm.devices["usb"] = TestDeviceCollection(
+            backend_vm=back_vm, devclass="usb"
+        )
+        front.devices["usb"] = TestDeviceCollection(
+            backend_vm=front, devclass="usb"
+        )
+        dom0.devices["usb"] = TestDeviceCollection(
+            backend_vm=dom0, devclass="usb"
+        )
+
+        return back_vm, front
+
+    def test_010_on_qdb_change_multiple_assignments_including_full(self):
+        back, front = self.added_assign_setup()
+
+        exp_dev = qubesusbproxy.core3ext.USBDevice(back, "1-1")
+        full_assig = DeviceAssignment(
+            VirtualDevice(exp_dev.port, exp_dev.device_id),
+            mode="auto-attach",
+            options={"pid": "did"},
+        )
+        port_assign = DeviceAssignment(
+            VirtualDevice(exp_dev.port, "*"),
+            mode="auto-attach",
+            options={"pid": "any"},
+        )
+        dev_assign = DeviceAssignment(
+            VirtualDevice(
+                Port(exp_dev.backend_domain, "*", "usb"), exp_dev.device_id
+            ),
+            mode="auto-attach",
+            options={"any": "did"},
+        )
+
+        front.devices["usb"]._assigned.append(dev_assign)
+        front.devices["usb"]._assigned.append(port_assign)
+        front.devices["usb"]._assigned.append(full_assig)
+        back.devices["usb"]._exposed.append(
+            qubesusbproxy.core3ext.USBDevice(back, "1-1")
+        )
+
+        self.ext.attach_and_notify = Mock()
```

Comment:
You can't do that, use patch instead.
It doesn't looks like it, bue extension object actually is a singleton, and this will affect other tests in this process too.
And also, this should be AsyncMock (@fepitre already fixed it in #43).

Reply:
> extension object actually is a singleton

Ok, that explains a lot

Reply:
I haven't tried manually fixing this, but if I skip this test class, the other one passes in full, no more "Device connection failed" errors etc :)